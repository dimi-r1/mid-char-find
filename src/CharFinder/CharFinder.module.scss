.paper {
    display: flex;
    flex-direction: column;
    align-items: center;
    text-align: center;

    .orText {
        margin-top: var(--s-1);
        margin-bottom: var(--s-1);
    }

    .wordCard {
        min-width: var(--s-max-content-small);
        margin-top: var(--s-3);
        margin-bottom: var(--s-3);
    }
}

.middle {
    color: var(--c-green);
}
