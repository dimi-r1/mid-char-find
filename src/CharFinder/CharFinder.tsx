import {
    Button,
    Card,
    CardContent,
    CircularProgress,
    Paper,
    TextField,
    Typography,
} from "@mui/material";
import React, { useCallback, useEffect, useState } from "react";
import styles from "./CharFinder.module.scss";
import { extractWordParts, getRandomWord } from "./utilities/utilities";

const CharFinder = () => {
    const [textInput, setTextInput] = useState("");
    const [word, setWord] = useState("");
    const [loading, setLoading] = useState(false);

    const setRandomWord = () => {
        setLoading(true);

        getRandomWord().then((newWord) => {
            setLoading(false);
            setWord(newWord);
        });
    };

    // Set a random word on component mount.
    useEffect(() => {
        setRandomWord();
    }, []);

    // Set a random word when requested by the user and clear the text input.
    const handleFetch = useCallback(() => {
        setRandomWord();
        setTextInput("");
    }, []);

    // Set word from user input.
    const handleTextInput = useCallback(
        (e: React.ChangeEvent<HTMLTextAreaElement | HTMLInputElement>) => {
            const userInput = e.target.value.trim();

            setTextInput(userInput);
            if (userInput) {
                setWord(userInput);
                // If the user is clearing the input, set random word once.
            } else if (!(!userInput && !textInput)) {
                setRandomWord();
            }
        },
        [textInput]
    );

    const parts = extractWordParts(word);

    return (
        <Paper className={styles.paper} elevation={0}>
            <Typography className={styles.title} variant="h1">
                Middle character finder
            </Typography>
            <Typography className={styles.title} variant="subtitle1">
                Press the fetch button to retreive a new word or type in a word.
                The middle character(s) will appear in the card below.
            </Typography>
            <Card className={styles.wordCard} elevation={4}>
                <CardContent>
                    {loading ? (
                        <CircularProgress />
                    ) : (
                        <Typography variant="h2" component="p">
                            <span>{parts.left}</span>
                            <strong className={styles.middle}>
                                {parts.middle}
                            </strong>
                            <span>{parts.right}</span>
                        </Typography>
                    )}
                </CardContent>
            </Card>
            <Button variant="outlined" onClick={handleFetch}>
                Fetch random word
            </Button>
            <Typography
                className={styles.orText}
                variant="button"
                component="p"
            >
                OR
            </Typography>
            <TextField
                id="outlined-basic"
                label="Type your own here"
                variant="outlined"
                value={textInput}
                onChange={handleTextInput}
            />
        </Paper>
    );
};

export default CharFinder;
