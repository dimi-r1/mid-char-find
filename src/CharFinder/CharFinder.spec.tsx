import React from "react";
import { render, screen } from "@testing-library/react";
import userEvent from "@testing-library/user-event";
import "@testing-library/jest-dom";

import CharFinder from "./CharFinder";

const unmockedFetch = global.fetch;

beforeAll(() => {
    global.fetch = () =>
        // @ts-ignore
        Promise.resolve({
            json: () =>
                Promise.resolve([
                    {
                        definition:
                            "Calf's skin set beside a cow to make her give milk freely  ",
                        pronunciation: "Tulkshan",
                        word: "Tulchan",
                    },
                ]),
        });
});

afterAll(() => {
    global.fetch = unmockedFetch;
});

describe("CharFinder", () => {
    it("should render correctly", () => {
        const { container } = render(<CharFinder />);
        expect(container).toMatchSnapshot();
    });

    it("should show a default word on component mount", async () => {
        render(<CharFinder />);

        expect(await screen.findByText("tul")).toBeInTheDocument();
        expect(screen.getByText("c")).toBeInTheDocument();
        expect(screen.getByText("han")).toBeInTheDocument();
    });

    it("should update the word when the user clicks the fetch button", async () => {
        global.fetch = () =>
            // @ts-ignore
            Promise.resolve({
                json: () =>
                    Promise.resolve([
                        {
                            definition: "test",
                            pronunciation: "test",
                            word: "tester",
                        },
                    ]),
            });

        render(<CharFinder />);

        const button = await screen.findByRole("button", { name: /Fetch/ });

        userEvent.click(button);

        expect(await screen.findByText("te")).toBeInTheDocument();
        expect(screen.getByText("st")).toBeInTheDocument();
        expect(screen.getByText("er")).toBeInTheDocument();
    });

    it("should update the word when the user types in custom text", async () => {
        render(<CharFinder />);

        const textBox = await screen.findByRole("textbox", { name: /Type/ });

        userEvent.type(textBox, "washing");

        expect(await screen.findByText("was")).toBeInTheDocument();
        expect(screen.getByText("h")).toBeInTheDocument();
        expect(screen.getByText("ing")).toBeInTheDocument();
    });

    it("should should fetch a randm word when the user deletes the textbox content", async () => {
        render(<CharFinder />);

        const textBox = await screen.findByRole("textbox", { name: /Type/ });

        userEvent.type(textBox, "washing");

        expect(await screen.findByText("was")).toBeInTheDocument();
        expect(screen.getByText("h")).toBeInTheDocument();
        expect(screen.getByText("ing")).toBeInTheDocument();

        userEvent.clear(textBox);

        expect(await screen.findByText("te")).toBeInTheDocument();
        expect(screen.getByText("st")).toBeInTheDocument();
        expect(screen.getByText("er")).toBeInTheDocument();
    });
});
