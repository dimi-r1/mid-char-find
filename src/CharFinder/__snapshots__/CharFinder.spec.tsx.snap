// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CharFinder should render correctly 1`] = `
<div>
  <div
    class="MuiPaper-root MuiPaper-elevation MuiPaper-rounded MuiPaper-elevation0 paper css-ee20y-MuiPaper-root"
  >
    <h1
      class="MuiTypography-root MuiTypography-h1 title css-o2w69a-MuiTypography-root"
    >
      Middle character finder
    </h1>
    <h6
      class="MuiTypography-root MuiTypography-subtitle1 title css-10wpov9-MuiTypography-root"
    >
      Press the fetch button to retreive a new word or type in a word. The middle character(s) will appear in the card below.
    </h6>
    <div
      class="MuiPaper-root MuiPaper-elevation MuiPaper-rounded MuiPaper-elevation4 MuiCard-root wordCard css-pgkxs3-MuiPaper-root-MuiCard-root"
    >
      <div
        class="MuiCardContent-root css-46bh2p-MuiCardContent-root"
      >
        <span
          class="MuiCircularProgress-root MuiCircularProgress-indeterminate MuiCircularProgress-colorPrimary css-18lrjg1-MuiCircularProgress-root"
          role="progressbar"
          style="width: 40px; height: 40px;"
        >
          <svg
            class="MuiCircularProgress-svg css-1idz92c-MuiCircularProgress-svg"
            viewBox="22 22 44 44"
          >
            <circle
              class="MuiCircularProgress-circle MuiCircularProgress-circleIndeterminate css-176wh8e-MuiCircularProgress-circle"
              cx="44"
              cy="44"
              fill="none"
              r="20.2"
              stroke-width="3.6"
            />
          </svg>
        </span>
      </div>
    </div>
    <button
      class="MuiButton-root MuiButton-outlined MuiButton-outlinedPrimary MuiButton-sizeMedium MuiButton-outlinedSizeMedium MuiButtonBase-root  css-1rwt2y5-MuiButtonBase-root-MuiButton-root"
      tabindex="0"
      type="button"
    >
      Fetch random word
      <span
        class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
      />
    </button>
    <p
      class="MuiTypography-root MuiTypography-button orText css-1f0on15-MuiTypography-root"
    >
      OR
    </p>
    <div
      class="MuiFormControl-root MuiTextField-root css-1u3bzj6-MuiFormControl-root-MuiTextField-root"
    >
      <label
        class="MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-outlined MuiFormLabel-root MuiFormLabel-colorPrimary css-14s5rfu-MuiFormLabel-root-MuiInputLabel-root"
        data-shrink="false"
        for="outlined-basic"
        id="outlined-basic-label"
      >
        Type your own here
      </label>
      <div
        class="MuiOutlinedInput-root MuiInputBase-root MuiInputBase-colorPrimary MuiInputBase-formControl css-9ddj71-MuiInputBase-root-MuiOutlinedInput-root"
      >
        <input
          aria-invalid="false"
          class="MuiOutlinedInput-input MuiInputBase-input css-1t8l2tu-MuiInputBase-input-MuiOutlinedInput-input"
          id="outlined-basic"
          type="text"
          value=""
        />
        <fieldset
          aria-hidden="true"
          class="MuiOutlinedInput-notchedOutline css-1d3z3hw-MuiOutlinedInput-notchedOutline"
        >
          <legend
            class="css-173wfxe"
          >
            <span>
              Type your own here
            </span>
          </legend>
        </fieldset>
      </div>
    </div>
  </div>
</div>
`;
